File: packages/sdk/src/events.ts
/**
 * Platform real-time events via Socket.IO at `/api/v1/events`.
 *
 * The `MelayaEvents` class exposes typed subscription helpers that mirror what
 * the Melaya frontend subscribes to in `SocketContext.tsx`. The underlying
 * Socket.IO connection is opened lazily on the first subscription/join call —
 * constructing the client performs no network I/O and never keeps a one-shot
 * process alive.
 *
 * Because Socket.IO uses its own protocol, this module does NOT reuse the
 * raw `WebSocket` used by `stream.ts`. Instead it builds a minimal, zero-dep
 * client that speaks the Engine.IO/Socket.IO wire format for HTTP long-polling
 * (the universal transport) and upgrades to WebSocket when available. For
 * Node environments we recommend also passing `WebSocket` in `MelayaOptions`.
 *
 * Room semantics (mirroring the server):
 *   - `run:<runId>` — subscribe to events for a specific pipeline run
 *   - `project:<project>` — subscribe to all events in a project
 *   - `hitl:user:<userId>` — HITL approval events (joined automatically for
 *     the authenticated user by the server on connect)
 *
 * @example
 * ```ts
 * const evts = melaya.events;
 * evts.onRunUpdate("run-123", (e) => console.log(e.event_type, e.status));
 * evts.onHitlApproval((e) => console.log("HITL:", e.type, e.requestId));
 *
 * // Clean up
 * evts.leaveRun("run-123");
 * evts.close();
 * ```
 */

import type {
  HitlApprovalEvent,
  PipelineCrudEvent,
  RunInitPhaseEvent,
  RunPushEvent,
} from "./platform-types.js";
import type { WebSocketCtor } from "./stream.js";

type Listener<T> = (payload: T) => void;
type AnyListener = Listener<unknown>;

/** Minimal Engine.IO / Socket.IO v4 packet types we care about. */
const EIO_PING = "2";
const EIO_PONG = "3";
const EIO_MESSAGE = "4"; // followed by Socket.IO packet

const SIO_CONNECT = 0;
const SIO_EVENT = 2;

/** Engine.IO v4 record separator between batched polling packets. */
const EIO_RECORD_SEPARATOR = "\x1e";

/**
 * A thin, zero-dependency Socket.IO v4 client (Engine.IO v4 underneath).
 *
 * Transport strategy: on the first subscription/join call, complete the
 * Engine.IO handshake over HTTP long-poll (the spec-mandated upgrade path),
 * then upgrade to WebSocket. In environments without WebSocket the client
 * stays on long-poll: incoming batches are split on the 0x1e record
 * separator, and outbound packets (pongs, room joins) are POSTed back to the
 * polling endpoint with the session id.
 *
 * The client only implements the subset needed for the Melaya events surface:
 *   - connect (emit `40` connect packet with auth)
 *   - join rooms (emit `42["joinRunRoom", ...]` / `42["joinProjectRoom", ...]`)
 *   - receive events (parse `42[...]` packets)
 *   - disconnect / close
 */
export class MelayaEvents {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly WebSocketImpl: WebSocketCtor | undefined;
  private readonly fetchImpl: typeof fetch;

  private sid: string | null = null;
  private ws: import("./stream.js").WebSocketLike | null = null;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private closed = false;
  /** True once the first subscription/join has triggered a connect. */
  private started = false;
  /** Reconnect attempt count for exponential backoff (reset on successful connect). */
  private reconnectAttempt = 0;
  /** Outbound packets queued while the WebSocket upgrade is still connecting. */
  private pendingOut: string[] = [];

  /** event name → Set of listeners */
  private readonly eventListeners = new Map<string, Set<AnyListener>>();
  /** rooms currently joined (replayed after every reconnect) */
  private readonly joinedRooms = new Set<string>();
  /** room → listeners registered through that room (removed by leave*) */
  private readonly roomListeners = new Map<string, Set<{ event: string; listener: AnyListener }>>();

  constructor(opts: {
    baseUrl: string;
    apiKey: string;
    WebSocket?: WebSocketCtor;
    fetch: typeof fetch;
  }) {
    // Derive the Socket.IO URL: same origin, path=/api/v1/events
    const base = opts.baseUrl.replace(/\/$/, "");
    this.baseUrl = base;
    this.apiKey = opts.apiKey;
    this.WebSocketImpl = opts.WebSocket;
    this.fetchImpl = opts.fetch;
    // NOTE: no connect() here — the connection opens lazily on first use.
  }

  // ── Public subscription API ────────────────────────────────────────────────

  /**
   * Subscribe to run events for a specific pipeline run.
   * Automatically joins the `run:<runId>` Socket.IO room. Events that carry a
   * `runId` for a different run (e.g. when multiple run rooms are joined) are
   * filtered out before your callback fires.
   *
   * Returns an unsubscribe function.
   */
  onRunUpdate(runId: string, cb: Listener<RunPushEvent>): () => void {
    const filtered: Listener<RunPushEvent> = (e) => {
      if (e && typeof e === "object" && e.runId !== undefined && e.runId !== runId) return;
      cb(e);
    };
    return this.registerRoomListener(`run:${runId}`, "pushEvent", filtered as AnyListener);
  }

  /**
   * Subscribe to init-phase progress events for a run.
   * Automatically joins the `run:<runId>` room; events for other runs are
   * filtered out.
   *
   * Returns an unsubscribe function.
   */
  onInitPhase(runId: string, cb: Listener<RunInitPhaseEvent>): () => void {
    const filtered: Listener<RunInitPhaseEvent> = (e) => {
      if (e && typeof e === "object" && e.runId !== undefined && e.runId !== runId) return;
      cb(e);
    };
    return this.registerRoomListener(`run:${runId}`, "pushInitPhase", filtered as AnyListener);
  }

  /**
   * Subscribe to all events in a project (all runs, pipeline CRUD).
   * Automatically joins the `project:<project>` room. Events that carry a
   * `project` for a different project are filtered out.
   *
   * Returns an unsubscribe function.
   */
  onProjectEvent(
    project: string,
    cb: Listener<RunPushEvent | PipelineCrudEvent>,
  ): () => void {
    const filtered: Listener<RunPushEvent | PipelineCrudEvent> = (e) => {
      if (e && typeof e === "object" && e.project !== undefined && e.project !== project) return;
      cb(e);
    };
    return this.registerRoomListener(`project:${project}`, "pushEvent", filtered as AnyListener);
  }

  /**
   * Subscribe to HITL approval invalidation events.
   * The server automatically joins authenticated sockets to the
   * `hitl:user:<userId>` room on connect — no explicit join needed.
   *
   * Returns an unsubscribe function.
   */
  onHitlApproval(cb: Listener<HitlApprovalEvent>): () => void {
    return this.on("pushHitlApprovals", cb as AnyListener);
  }

  /** Subscribe to pipeline CRUD events within a project room. */
  onPipelineCreated(project: string, cb: Listener<PipelineCrudEvent>): () => void {
    return this.subscribeProjectCrud(project, "pipelineCreated", cb);
  }

  /** Subscribe to pipeline update events within a project room. */
  onPipelineUpdated(project: string, cb: Listener<PipelineCrudEvent>): () => void {
    return this.subscribeProjectCrud(project, "pipelineUpdated", cb);
  }

  /** Subscribe to pipeline deletion events within a project room. */
  onPipelineDeleted(project: string, cb: Listener<PipelineCrudEvent>): () => void {
    return this.subscribeProjectCrud(project, "pipelineDeleted", cb);
  }

  /**
   * Stop listening to a specific run (leave the run room and remove its listeners).
   * Listeners registered for other events/rooms are unaffected.
   */
  leaveRun(runId: string): void {
    this.leaveRoom(`run:${runId}`);
  }

  /** Leave a project room and remove the listeners registered through it. */
  leaveProject(project: string): void {
    this.leaveRoom(`project:${project}`);
  }

  /** Close the Socket.IO connection and release all listeners. */
  close(): void {
    this.closed = true;
    if (this.reconnectTimer) { clearTimeout(this.reconnectTimer); this.reconnectTimer = null; }
    this.ws?.close();
    this.ws = null;
    this.sid = null;
    this.pendingOut = [];
    this.eventListeners.clear();
    this.joinedRooms.clear();
    this.roomListeners.clear();
  }

  /**
   * Compute a bounded exponential backoff delay with jitter.
   * Caps at 30 000 ms; resets reconnectAttempt on successful connect.
   */
  private nextReconnectDelay(): number {
    const base = Math.min(1000 * 2 ** this.reconnectAttempt, 30_000);
    // ±20% jitter
    const jitter = base * 0.2 * (Math.random() * 2 - 1);
    this.reconnectAttempt = Math.min(this.reconnectAttempt + 1, 10);
    return Math.max(500, base + jitter);
  }

  // ── Internal helpers ───────────────────────────────────────────────────────

  /** Lazily open the connection on the first subscription/join. */
  private ensureConnected(): void {
    if (this.started || this.closed) return;
    this.started = true;
    void this.connect();
  }

  /** Register a raw Socket.IO event listener. Returns an unsubscribe fn. */
  private on(event: string, cb: AnyListener): () => void {
    this.ensureConnected();
    if (!this.eventListeners.has(event)) {
      this.eventListeners.set(event, new Set());
    }
    this.eventListeners.get(event)!.add(cb);
    return () => {
      this.eventListeners.get(event)?.delete(cb);
    };
  }

  /**
   * Join a room and register a listener that belongs to it, so a later
   * `leaveRun()` / `leaveProject()` can remove exactly this listener.
   */
  private registerRoomListener(room: string, event: string, listener: AnyListener): () => void {
    this.joinRoom(room);
    const off = this.on(event, listener);
    let set = this.roomListeners.get(room);
    if (!set) { set = new Set(); this.roomListeners.set(room, set); }
    const entry = { event, listener };
    set.add(entry);
    return () => {
      off();
      this.roomListeners.get(room)?.delete(entry);
    };
  }

  /** Shared body of the onPipelineCreated/Updated/Deleted subscriptions. */
  private subscribeProjectCrud(
    project: string,
    event: string,
    cb: Listener<PipelineCrudEvent>,
  ): () => void {
    const filtered: Listener<PipelineCrudEvent> = (e) => {
      if (e && typeof e === "object" && e.project !== undefined && e.project !== project) return;
      cb(e);
    };
    return this.registerRoomListener(`project:${project}`, event, filtered as AnyListener);
  }

  private emit(event: string, payload: unknown): void {
    const listeners = this.eventListeners.get(event);
    if (!listeners) return;
    for (const l of [...listeners]) {
      try { l(payload); } catch { /* listener errors must not crash the client */ }
    }
  }

  /** Handle a decoded Socket.IO event packet. */
  private handleSioEvent(args: unknown[]): void {
    if (!Array.isArray(args) || args.length < 2) return;
    const [event, payload] = args as [string, unknown];
    this.emit(event, payload);
  }

  /** Decode a raw Engine.IO packet string and dispatch. */
  private handlePacket(raw: string): void {
    if (!raw) return;
    const type = raw[0];
    if (type === EIO_PING) {
      // Engine.IO heartbeat — answer promptly with a pong on whatever
      // transport is active (WebSocket frame or polling POST).
      this.sendPacket(EIO_PONG);
      return;
    }
    if (type !== EIO_MESSAGE) return;
    // Socket.IO packet starts at index 1
    const sioRaw = raw.slice(1);
    const sioType = parseInt(sioRaw[0], 10);
    if (sioType !== SIO_EVENT) return;
    try {
      const args = JSON.parse(sioRaw.slice(1)) as unknown[];
      this.handleSioEvent(args);
    } catch {
      // ignore malformed packets
    }
  }

  // ── Connection lifecycle ───────────────────────────────────────────────────

  private pollUrl(params: Record<string, string> = {}): string {
    const u = new URL(`${this.baseUrl}/`);
    // Engine.IO endpoint is the Socket.IO path
    u.pathname = "/api/v1/events/";
    u.searchParams.set("EIO", "4");
    u.searchParams.set("transport", "polling");
    if (this.sid) u.searchParams.set("sid", this.sid);
    for (const [k, v] of Object.entries(params)) u.searchParams.set(k, v);
    return u.toString();
  }

  private wsUrl(): string {
    const u = new URL(`${this.baseUrl}/`);
    u.protocol = u.protocol === "https:" ? "wss:" : "ws:";
    u.pathname = "/api/v1/events/";
    u.searchParams.set("EIO", "4");
    u.searchParams.set("transport", "websocket");
    if (this.sid) u.searchParams.set("sid", this.sid);
    return u.toString();
  }

  private wsSend(data: string): void {
    if (this.ws && (this.ws.readyState === 1 /* OPEN */)) {
      try { (this.ws as unknown as { send: (d: string) => void }).send(data); } catch { /* ignore */ }
    }
  }

  /**
   * Send an Engine.IO packet over whatever transport is active:
   *   - WebSocket open → send as a frame
   *   - WebSocket still connecting → queue, flushed on open
   *   - long-poll → POST to the polling endpoint with the session id
   */
  private sendPacket(packet: string): void {
    if (this.ws) {
      if (this.ws.readyState === 1 /* OPEN */) {
        this.wsSend(packet);
      } else {
        this.pendingOut.push(packet);
      }
      return;
    }
    if (this.sid) {
      void this.postPacket(packet);
    }
    // Not connected yet: room joins are replayed after connect; anything else
    // (pongs) is meaningless without a session and is dropped.
  }

  /** POST one outbound packet to the long-poll endpoint. */
  private async postPacket(packet: string): Promise<void> {
    try {
      const res = await this.fetchImpl(this.pollUrl(), {
        method: "POST",
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          "Content-Type": "text/plain;charset=UTF-8",
        },
        body: packet,
      });
      if (res.status === 400 && !this.closed && !this.ws) {
        // "Session ID unknown" — the session is dead; full re-handshake.
        this.sid = null;
        this.scheduleReconnect();
      }
    } catch {
      /* transient network error; the poll loop detects dead sessions */
    }
  }

  /** Build the Socket.IO connect packet, including the Bearer token for auth. */
  private sioConnectPacket(): string {
    // "40" = EIO message (4) + SIO connect (0) + optional JSON data
    const authData = JSON.stringify({ token: this.apiKey });
    return `${EIO_MESSAGE}${SIO_CONNECT}${authData}`;
  }

  /** Emit a Socket.IO event packet over whatever transport is active. */
  private sioEmit(event: string, ...args: unknown[]): void {
    const packet = `${EIO_MESSAGE}${SIO_EVENT}${JSON.stringify([event, ...args])}`;
    this.sendPacket(packet);
  }

  /**
   * Map a logical room string to the correct server emit + arg.
   * - "run:<id>"      → joinRunRoom(<id>)
   * - "project:<p>"   → joinProjectRoom(<p>)
   * - else            → joinRunRoom(<room>) as a safe fallback
   */
  private emitJoin(room: string): void {
    if (room.startsWith("run:")) {
      this.sioEmit("joinRunRoom", room.slice(4));
    } else if (room.startsWith("project:")) {
      this.sioEmit("joinProjectRoom", room.slice(8));
    } else {
      // Unknown room prefix — best-effort fallback
      this.sioEmit("joinRunRoom", room);
    }
  }

  private joinRoom(room: string): void {
    this.ensureConnected();
    if (this.joinedRooms.has(room)) return;
    this.joinedRooms.add(room);
    if (this.sid) {
      this.emitJoin(room);
    }
    // if not connected yet, joinedRooms will be replayed in replayJoins()
  }

  private leaveRoom(room: string): void {
    // Remove the listeners registered through this room so callbacks stop
    // firing, and drop the room from the rejoin-on-reconnect list.
    const set = this.roomListeners.get(room);
    if (set) {
      for (const { event, listener } of set) {
        this.eventListeners.get(event)?.delete(listener);
      }
      this.roomListeners.delete(room);
    }
    this.joinedRooms.delete(room);
    if (this.sid) {
      // Server expects leaveRoom with the full room string ("run:<id>" / "project:<p>")
      this.sioEmit("leaveRoom", room);
    }
  }

  /** Replay all pending room joins after a successful connect. */
  private replayJoins(): void {
    for (const room of this.joinedRooms) {
      this.emitJoin(room);
    }
  }

  /**
   * Schedule a reconnect with bounded exponential backoff. The timer is
   * unref()ed (where supported) so a pending reconnect never keeps a Node
   * process alive on its own.
   */
  private scheduleReconnect(): void {
    if (this.closed || this.reconnectTimer) return;
    const t = setTimeout(() => {
      this.reconnectTimer = null;
      void this.connect();
    }, this.nextReconnectDelay());
    (t as unknown as { unref?: () => void }).unref?.();
    this.reconnectTimer = t;
  }

  private async connect(): Promise<void> {
    if (this.closed) return;
    try {
      // Step 1: Engine.IO handshake via HTTP polling
      const res = await this.fetchImpl(this.pollUrl(), {
        headers: { Authorization: `Bearer ${this.apiKey}` },
      });
      const text = await res.text();
      // EIO4 open packet format: "0{...json...}"
      const jsonStart = text.indexOf("{");
      if (!res.ok || jsonStart === -1) throw new Error("MelayaEvents: unexpected EIO handshake response");
      const openData = JSON.parse(
        text.slice(jsonStart).split(EIO_RECORD_SEPARATOR)[0],
      ) as { sid: string; pingInterval?: number; pingTimeout?: number };
      this.sid = openData.sid;

      // Step 2: Send Socket.IO connect packet with auth via polling POST
      const post = await this.fetchImpl(this.pollUrl(), {
        method: "POST",
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          "Content-Type": "text/plain;charset=UTF-8",
        },
        body: this.sioConnectPacket(),
      });
      if (!post.ok) throw new Error("MelayaEvents: Socket.IO connect packet rejected");

      this.reconnectAttempt = 0; // successful connect — reset backoff

      // Step 3: Upgrade to WebSocket if available; otherwise stay on long-poll
      if (this.WebSocketImpl && !this.closed) {
        this.upgradeToWs(); // joins are replayed once the socket opens
      } else if (!this.closed) {
        this.replayJoins();
        void this.pollLoop();
      }
    } catch {
      // Retry with bounded exponential back-off
      if (!this.closed) {
        this.scheduleReconnect();
      }
    }
  }

  private upgradeToWs(): void {
    if (!this.WebSocketImpl || this.closed) return;
    const ws = new this.WebSocketImpl(this.wsUrl());
    this.ws = ws;
    ws.addEventListener("message", (ev: unknown) => {
      const raw = (ev as { data?: unknown }).data;
      const text = typeof raw === "string" ? raw
        : raw instanceof ArrayBuffer ? new TextDecoder().decode(raw) : String(raw);
      this.handlePacket(text);
    });
    ws.addEventListener("close", () => {
      this.ws = null;
      this.pendingOut = [];
      if (!this.closed) {
        // reconnect with bounded exponential backoff
        this.sid = null;
        this.scheduleReconnect();
      }
    });
    ws.addEventListener("error", () => {
      /* close will follow */
    });
    ws.addEventListener("open", () => {
      // Engine.IO WebSocket upgrade probe
      this.wsSend("2probe");
      this.wsSend("5"); // upgrade packet
      // Flush anything queued while connecting, then (re)join our rooms.
      const queued = this.pendingOut;
      this.pendingOut = [];
      for (const p of queued) this.wsSend(p);
      this.replayJoins();
    });
  }

  /**
   * Long-poll receive loop (used when no WebSocket implementation is
   * available). Each GET is held by the server until packets are ready, so
   * the loop re-polls immediately. An HTTP 400 ("Session ID unknown") means
   * the session died server-side → full re-handshake.
   */
  private async pollLoop(): Promise<void> {
    while (!this.closed && !this.ws && this.sid) {
      try {
        const res = await this.fetchImpl(this.pollUrl(), {
          headers: { Authorization: `Bearer ${this.apiKey}` },
        });
        if (res.status === 400) {
          this.sid = null;
          this.scheduleReconnect();
          return;
        }
        const text = await res.text();
        this.parsePollResponse(text);
      } catch {
        if (!this.closed) {
          this.sid = null;
          this.scheduleReconnect();
        }
        return;
      }
    }
  }

  private parsePollResponse(text: string): void {
    // Engine.IO v4 batches packets separated by 0x1e (record separator).
    for (const packet of text.split(EIO_RECORD_SEPARATOR)) {
      if (packet) this.handlePacket(packet);
    }
  }
}

File: packages/sdk-ruby/lib/melaya/events.rb
# frozen_string_literal: true

require "uri"
require "net/http"
require "json"
require "openssl"
require "set"
require "thread"

require_relative "errors"

module Melaya
  # Platform real-time events via Socket.IO v4 at /api/v1/events.
  #
  # Implements the Engine.IO v4 handshake and long-poll transport (packets
  # batched with the 0x1e record separator; server pings answered with a
  # POSTed pong) and the Socket.IO v4 packet framing, mirroring the TS
  # events.ts client.
  #
  # Room semantics (matching the server):
  #   - run:<runId>          — pipeline run events
  #   - project:<project>    — all events in a project (runs + CRUD)
  #   - hitl:user:<userId>   — HITL approval events (joined automatically by server)
  #
  # Thread safety: all public methods are safe to call from any thread.
  # Callbacks are invoked on a dedicated background reader thread — do not
  # block in a callback.
  #
  # @example
  #   events = Melaya::Events.new(api_key: ENV["MELAYA_API_KEY"])
  #
  #   # Subscribe to a run
  #   unsub = events.on_run_update("run-123") { |e| puts e["event_type"] }
  #
  #   # Subscribe to HITL approvals
  #   events.on_hitl_approval { |e| puts "HITL: #{e["type"]}" }
  #
  #   # Clean up
  #   unsub.call          # remove one subscription
  #   events.leave_run("run-123")
  #   events.close
  #
  # NOTE: Requires Ruby standard library 'net/http' with persistent HTTP/1.1.
  # For production use the connection is kept alive between polls.
  class Events
    # Engine.IO v4 packet type characters
    EIO_OPEN    = "0"
    EIO_CLOSE   = "1"
    EIO_PING    = "2"
    EIO_PONG    = "3"
    EIO_MESSAGE = "4"
    EIO_UPGRADE = "5"
    EIO_NOOP    = "6"

    # Socket.IO v4 packet types (numeric, follows the EIO_MESSAGE prefix)
    SIO_CONNECT    = 0
    SIO_DISCONNECT = 1
    SIO_EVENT      = 2
    SIO_ACK        = 3
    SIO_ERROR      = 4

    RECONNECT_BASE_DELAY   = 2  # seconds — first reconnect wait
    RECONNECT_MAX_DELAY    = 30 # seconds — cap for exponential backoff

    # @param api_key [String] mk_* key
    # @param base_url [String]
    # @param verify_ssl [Boolean]
    def initialize(api_key:, base_url: HttpClient::DEFAULT_BASE_URL, verify_ssl: true)
      raise ArgumentError, "Melaya: TLS certificate verification cannot be disabled." unless verify_ssl
      @_tok       = api_key.freeze
      @base_url   = base_url.sub(/\/$/, "")
      @verify_ssl = true

      @sid            = nil
      @closed         = false
      @listeners      = Hash.new { |h, k| h[k] = [] }  # event → [callable]
      @joined_rooms   = Set.new                        # rejoin-on-reconnect list
      @room_listeners = Hash.new { |h, k| h[k] = [] }  # room  → [[event, callable]]
      @mutex          = Mutex.new

      _connect_async
    end

    # ── Subscription helpers ───────────────────────────────────────────────────

    # Subscribe to run events for a specific pipeline run.
    # Joins the run:<runId> Socket.IO room automatically. When the payload
    # carries a +runId+, events for other runs are filtered out so multiple
    # per-run subscriptions never receive cross-room events.
    # @param run_id [String]
    # @yield [Hash] run push event
    # @return [Proc] call to unsubscribe
    def on_run_update(run_id, &block)
      join_room("run:#{run_id}")
      on("pushEvent", room: "run:#{run_id}", &run_filter(run_id, block))
    end

    # Subscribe to init-phase progress events for a run.
    # @param run_id [String]
    # @yield [Hash]
    # @return [Proc]
    def on_init_phase(run_id, &block)
      join_room("run:#{run_id}")
      on("pushInitPhase", room: "run:#{run_id}", &run_filter(run_id, block))
    end

    # Subscribe to all events in a project (runs + pipeline CRUD). When the
    # payload carries a +projectId+/+project+, events for other projects are
    # filtered out.
    # @param project [String]
    # @yield [Hash]
    # @return [Proc]
    def on_project_event(project, &block)
      join_room("project:#{project}")
      on("pushEvent", room: "project:#{project}", &project_filter(project, block))
    end

    # Subscribe to HITL approval invalidation events.
    # The server joins authenticated sockets to the hitl:user:<userId> room
    # on connect — no explicit join needed.
    # @yield [Hash]
    # @return [Proc]
    def on_hitl_approval(&block)
      on("pushHitlApprovals", &block)
    end

    # Subscribe to pipeline-created events within a project room.
    # @param project [String]
    # @yield [Hash]
    # @return [Proc]
    def on_pipeline_created(project, &block)
      join_room("project:#{project}")
      on("pipelineCreated", room: "project:#{project}", &project_filter(project, block))
    end

    # Subscribe to pipeline-updated events within a project room.
    # @param project [String]
    # @yield [Hash]
    # @return [Proc]
    def on_pipeline_updated(project, &block)
      join_room("project:#{project}")
      on("pipelineUpdated", room: "project:#{project}", &project_filter(project, block))
    end

    # Subscribe to pipeline-deleted events within a project room.
    # @param project [String]
    # @yield [Hash]
    # @return [Proc]
    def on_pipeline_deleted(project, &block)
      join_room("project:#{project}")
      on("pipelineDeleted", room: "project:#{project}", &project_filter(project, block))
    end

    # Leave a run room, remove its listeners, and drop it from the
    # rejoin-on-reconnect list.
    # @param run_id [String]
    def leave_run(run_id)
      leave_room("run:#{run_id}")
    end

    # Leave a project room, remove its listeners, and drop it from the
    # rejoin-on-reconnect list.
    # @param project [String]
    def leave_project(project)
      leave_room("project:#{project}")
    end

    # Close the Socket.IO connection and release all listeners.
    def close
      @mutex.synchronize do
        @closed = true
        @joined_rooms.clear
        @room_listeners.clear
        @listeners.clear
      end
      @poll_thread&.kill
    end

    private

    # ── Internal event bus ─────────────────────────────────────────────────────

    # Register a listener. When +room:+ is given the listener is also indexed
    # by room so leave_room can remove it.
    def on(event, room: nil, &block)
      @mutex.synchronize do
        @listeners[event] << block
        @room_listeners[room] << [event, block] if room
      end
      lambda do
        @mutex.synchronize do
          @listeners[event].delete(block)
          @room_listeners[room].delete([event, block]) if room
        end
      end
    end

    # Wrap +block+ so it only fires for the given run. Payloads that carry a
    # runId (RunPushEvent) are matched exactly; payloads without one pass
    # through unchanged.
    def run_filter(run_id, block)
      lambda do |payload|
        rid = payload.is_a?(Hash) ? (payload["runId"] || payload["run_id"]) : nil
        block.call(payload) if rid.nil? || rid.to_s == run_id.to_s
      end
    end

    # Wrap +block+ so it only fires for the given project. Payloads that carry
    # a projectId/project are matched exactly; payloads without one pass
    # through unchanged.
    def project_filter(project, block)
      lambda do |payload|
        pid = payload.is_a?(Hash) ? (payload["projectId"] || payload["project"]) : nil
        block.call(payload) if pid.nil? || pid.to_s == project.to_s
      end
    end

    def emit(event, payload)
      listeners = @mutex.synchronize { @listeners[event].dup }
      listeners.each do |l|
        l.call(payload) rescue nil
      end
    end

    # ── Room management ────────────────────────────────────────────────────────

    # Map a logical room string to the correct Socket.IO emit call.
    #
    # Server contract (/api/v1/events, server/src/index.ts):
    #   "run:<id>"      → emit "joinRunRoom",     bare runId (without "run:" prefix)
    #   "project:<name>"→ emit "joinProjectRoom", bare project name
    #   anything else   → emit "joinRunRoom" with the full string (safe fallback)
    #
    # leave(room) always emits "leaveRoom" with the FULL room string.
    def join_room(room)
      already = @mutex.synchronize do
        next true if @joined_rooms.include?(room)
        @joined_rooms.add(room)
        false
      end
      return if already
      _emit_join(room) if @sid
    end

    # Remove the room from the rejoin-on-reconnect list AND detach every
    # listener that was registered for it, then tell the server to leave.
    def leave_room(room)
      @mutex.synchronize do
        @joined_rooms.delete(room)
        (@room_listeners.delete(room) || []).each do |(event, block)|
          @listeners[event].delete(block)
        end
      end
      sio_emit("leaveRoom", room) if @sid
    end

    def replay_joins
      rooms = @mutex.synchronize { @joined_rooms.to_a }
      rooms.each { |room| _emit_join(room) }
    end

    # Emit the correct server event for a join, mapping the room string to the
    # appropriate event name and argument per the server protocol.
    def _emit_join(room)
      if room.start_with?("run:")
        sio_emit("joinRunRoom", room[4..])
      elsif room.start_with?("project:")
        sio_emit("joinProjectRoom", room[8..])
      else
        # Safe fallback for unknown room prefixes (e.g. future room types).
        sio_emit("joinRunRoom", room)
      end
    end

    # ── Packet dispatch ────────────────────────────────────────────────────────

    def handle_packet(raw)
      return if raw.nil? || raw.empty?
      type = raw[0]
      case type
      when EIO_PING
        poll_post(EIO_PONG)
      when EIO_MESSAGE
        handle_sio_packet(raw[1..])
      end
    end

    def handle_sio_packet(sio_raw)
      return if sio_raw.nil? || sio_raw.empty?
      sio_type = sio_raw[0].to_i
      return unless sio_type == SIO_EVENT
      begin
        args = JSON.parse(sio_raw[1..])
        return unless args.is_a?(Array) && args.size >= 2
        event, payload = args[0], args[1]
        emit(event, payload)
      rescue JSON::ParserError
        # ignore malformed packets
      end
    end

    # Engine.IO v4 polling batches packets separated by the 0x1e record
    # separator. Split on it and dispatch each packet individually.
    def parse_poll_response(text)
      text.split("\x1e").each { |packet| handle_packet(packet) }
    end

    # ── Socket.IO framing ──────────────────────────────────────────────────────

    def sio_connect_packet
      auth = JSON.generate({ "token" => @_tok })
      "#{EIO_MESSAGE}#{SIO_CONNECT}#{auth}"
    end

    def sio_emit(event, *args)
      packet = "#{EIO_MESSAGE}#{SIO_EVENT}#{JSON.generate([event, *args])}"
      poll_post(packet)
    end

    # ── HTTP polling transport ─────────────────────────────────────────────────

    def poll_uri(extra = {})
      params = {
        "EIO"       => "4",
        "transport" => "polling"
      }
      params["sid"] = @sid if @sid
      params.merge!(extra)
      uri = URI.parse("#{@base_url}/api/v1/events/")
      uri.query = URI.encode_www_form(params)
      uri
    end

    def http_for(uri)
      h = Net::HTTP.new(uri.host, uri.port)
      h.use_ssl     = uri.scheme == "https"
      h.verify_mode = OpenSSL::SSL::VERIFY_PEER
      h.open_timeout = 15
      h.read_timeout = 60
      h
    end

    def poll_get
      uri = poll_uri
      req = Net::HTTP::Get.new(uri)
      req["Authorization"] = "Bearer #{@_tok}"
      resp = http_for(uri).request(req)
      code = resp.code.to_i
      if code >= 400
        # HTTP 400 = "Session ID unknown": the sid is dead. Raising here makes
        # _connect_loop perform a full re-handshake (and replay room joins)
        # instead of silently polling a dead session forever.
        raise MelayaError.new("MelayaEvents: poll failed (HTTP #{code})", status: code)
      end
      resp.body.to_s
    end

    def poll_post(body_str)
      uri = poll_uri
      req = Net::HTTP::Post.new(uri)
      req["Authorization"] = "Bearer #{@_tok}"
      req["Content-Type"]  = "text/plain;charset=UTF-8"
      req.body = body_str
      http_for(uri).request(req)
    end

    # ── Connection lifecycle ───────────────────────────────────────────────────

    def _connect_async
      @poll_thread = Thread.new do
        Thread.current.abort_on_exception = false
        _connect_loop
      end
      @poll_thread.name = "melaya-events-poll"
    end

    def _connect_loop
      reconnect_attempt = 0
      loop do
        break if @mutex.synchronize { @closed }
        begin
          _handshake
          reconnect_attempt = 0  # reset on successful connection
          _poll_loop
        rescue StandardError
          # reconnect after bounded exponential backoff with ±25 % jitter
        end
        break if @mutex.synchronize { @closed }
        reconnect_attempt += 1
        base   = [RECONNECT_BASE_DELAY * (2**(reconnect_attempt - 1)), RECONNECT_MAX_DELAY].min.to_f
        jitter = base * 0.25 * (rand - 0.5) * 2
        sleep([base + jitter, 0.5].max)
        @sid = nil
      end
    end

    def _handshake
      # Step 1: Engine.IO open handshake
      text = poll_get
      json_start = text.index("{")
      raise MelayaError.new("MelayaEvents: unexpected EIO handshake", status: 0) if json_start.nil?
      data = JSON.parse(text[json_start..])
      @sid = data["sid"]

      # Step 2: Socket.IO connect packet with auth
      poll_post(sio_connect_packet)

      # Replay room joins
      replay_joins
    end

    # Continuous long-poll: the GET hangs server-side until data (or a ping)
    # arrives, so we re-issue it immediately — never sleep between polls, or
    # events queue up (and pings go unanswered) for the whole interval.
    # Errors (including dead-session HTTP 400) propagate to _connect_loop,
    # which backs off and re-handshakes.
    def _poll_loop
      loop do
        break if @mutex.synchronize { @closed }
        text = poll_get
        parse_poll_response(text) unless text.empty?
      end
    end
  end
end
